File: book/python_algorithm_interview/Q/09.py
from typing import List

# brute force


def three_num_for_zero(nums: List[int]) -> List[List[int]]:
    result = []
    n = len(nums)

    for i in range(n-2):
        for j in range(i+1, n-1):
            for k in range(j+1, n):
                if nums[i] + nums[j] + nums[k] == 0:
                    key_nums = sorted([nums[i], nums[j], nums[k]])
                    if key_nums not in result:
                        result.append(key_nums)

    return result


def threeSum(nums: List[int]) -> List[List[int]]:
    results = []
    nums.sort()

    for i in range(len(nums) - 2):
        if i > 0 and nums[i] == nums[i-1]:
            continue
        for j in range(i+1, len(nums) - 1):
            if j > i + 1 and nums[j] == nums[j-1]:
                continue
            for k in range(j+1, len(nums)):
                if k > j + 1 and nums[k] == nums[k-1]:
                    continue
                if nums[i] + nums[j] + nums[k] == 0:
                    results.append([nums[i], nums[j], nums[k]])

    return results

# two pointer


def threeSumTwoPointer(nums: List[int]) -> List[List[int]]:
    results = []
    nums.sort()

    for i in range(len(nums) - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        left, right = i + 1, len(nums) - 1

        while left < right:
            sum = nums[i] + nums[left] + nums[right]
            if sum < 0:
                left += 1
            elif sum > 0:
                right -= 1
            else:
                results.append([nums[i], nums[left], nums[right]])

                while left < right and nums[left] == nums[left + 1]:
                    left += 1
                while left < right and nums[right] == nums[right - 1]:
                    right += 1
                left += 1
                right -= 1

    return results


print(threeSumTwoPointer([-1, 0, 1, 2, -1, -4]))

def threeSumTest(nums: List[int]) -> List[List[int]]:
    results = []
    nums.sort()

    for i, v in enumerate(nums):
        # 첫번째 자리의 중복을 피하기 위해 이전에 검사한 값과 같을 경우 건너뛴다.
        if i > 0 and nums[i-1] == v:
            continue

        l, r = i+1, len(nums) - 1

        while l < r:
            threeSum = v + nums[l] + nums[r]

            # 합이 0 보다 작으면 l 을 올려서 합을 키운다.
            if threeSum < 0:
                l += 1
            # 합이 0 보다 크면 r 을 낮춰서 합을 줄인다.
            elif threeSum > 0:
                r -= 1
            else:
                # 합이 0 일 경우 결과 배열에 담는다.
                results.append([v, nums[l], nums[r]])

                # l, r 값이 중복일 경우 다시 검사할 필요 없기 때문에 while 을 통해 건너뛴다.
                while l < r and nums[l] == nums[l-1]:
                    l += 1
                while l < r and nums[r] == nums[r+1]:
                    r -= 1
                
                # 어차피 0 을 만들어야하고, l, r 중 하나에 따라 다른 한쪽도 고정되기 때문에 둘 다 옮겨줘도 된다.
                l += 1
                r -= 1
    
    return results

File: achieve/book/python_algorithm_interview/Q/09.ts
// not contain duplicate triplets.

// 더해서 세개가 돼야함.
// 알고리즘은? =>
// - 배열을 순회하면서 합이 0인 값 세개를 찾아야함.
// - 그러려면, 정렬해서 투포인터 사용. 정렬하고 가장 앞에 값을 curr 로 잡고, 나머지 구간을 투포인터 순회.
// - 위의 방식으로 순회하면서 가능성 있는 구간을 다 돌면 리턴.
// - 정렬했을 때, 앞의 수가 이미 0보다 크면 가망 없음. 리턴.

function threeSumBrouteForce(nums: number[]): number[][] {
  const results: number[][] = [];

  for (let i = 0; i < nums.length - 2; i++) {
    for (let j = i + 1; j < nums.length - 1; j++) {
      for (let k = j + 1; k < nums.length; k++) {
        const sum = nums[i] + nums[j] + nums[k];
        if (sum === 0) {
          const target = [nums[i], nums[j], nums[k]].sort();
          const isDup = results.some((v) => v.toString() === target.toString());
          if (!isDup) {
            results.push(target);
          }
        }
      }
    }
  }

  return results;
}

function threeSum(nums: number[]): number[][] {
  const results: number[][] = [];

  // sort
  nums.sort((a, b) => a - b);

  if (nums[0] > 0) return results;

  // loop with curr
  for (const [curr, _] of nums.entries()) {
    if (curr > 0 && nums[curr] === nums[curr - 1]) continue;

    // two pointer
    let left = curr + 1;
    let right = nums.length - 1;

    while (left < right) {
      const sum = nums[curr] + nums[left] + nums[right];

      if (sum === 0) {
        results.push([nums[curr], nums[left], nums[right]]);
        while (left < right && nums[left] === nums[left + 1]) left += 1;
        while (left < right && nums[right] === nums[right - 1]) right -= 1;
        // 얼핏 생각해보면 left 또는 right 둘 중 하나만 이동해야 하는 게 아닌가 싶지만, 어차피
        // sum=0인 상황이므로 어느 한쪽만 이동하는 경우는 불가능하다.
        left += 1;
        right -= 1;
      } else if (sum < 0) {
        left += 1;
      } else if (sum > 0) {
        right -= 1;
      }
    }
  }

  return [...new Set(results.map((item) => JSON.stringify(item)))].map((item) =>
    JSON.parse(item)
  );
}

const nums = [-1, 0, 1, 2, -1, -4];
// [-3,-1,4] 있는지 확인.
console.log(threeSum(nums));
//=> [[-1,-1,2],[-1,0,1]]
